storage: rationalize server-side refreshes and fix bugs

andreimatei opened the pull request:
Before this patch, we had several issues due to the server erroneously
considering that it's OK to commit a transaction at a bumped timestamp.

One of the issues was a lost update: a CPut could erroneously succeed
even though there's been a more recent write. This was caused by faulty
code in evaluateBatch() that was thinking that, just because an EndTxn
claimed to have been able to commit a transaction, that means that any
WriteTooOldError encountered previously by the batch was safe to
discard. An EndTxn might consider that it can commit even if there had
been previous write too old conditions if the NoRefreshSpans flag is
set. The problems is that a CPut that had returned a WriteTooOldError
also evaluated at the wrong read timestamp, and so its evaluation can't
be relied on.

Another issue is that, when the EndTxn code mentioned above considers
that it's safe to commit at a bumped timestamp, it doesn't take into
considerations that the EndTxn's batch might have performed reads (other
than CPuts)  that have been evaluated at a lower timestamp. This can
happen, for example in the following scenario: - a txn sends a Put which
gets bumped by the ts cache - the txn then sends a Scan + EndTxn. The
scan gets evaluated at the original timestamp, but then we commit at a
bumped one because the NoRefreshSpans flag is set.

The patch fixes the bugs by reworking how evaluation takes advantage of
the fact that some requests have flexible timestamps. EndTxn no longer
is in the business of committing at bumped timestamps, and its code is
thus simplified. Instead, the replica's "local retries" loop takes over.
The replica already had code handling non-transactional batches that
evaluated them repeatedly in case of WriteTooOldErrors. This patch
rationalizes and expands this code to deal with transactional batches
too, and with pushes besides WriteTooOldErrors. This reevaluation loop
now handles the cases in which the EndTxn used to bump the commit
timestamp.

The patch also fixes a third bug: the logic evaluateBatch() for
resetting the WriteTooOld state after a successful EndTransaction was
ignoring the STAGING state, meaning that the server would return a
WriteTooOldError even though the transaction was committed. I'm not sure
if this had dramatic consequences or was benign...

Fixes #42849

Release note (bug fix): A bug causing lost update transaction anomalies
was fixed.

andreimatei committed:
storage: rationalize server-side refreshes and fix bugs

Before this patch, we had several issues due to the server erroneously
considering that it's OK to commit a transaction at a bumped timestamp.

One of the issues was a lost update: a CPut could erroneously succeed
even though there's been a more recent write. This was caused by faulty
code in evaluateBatch() that was thinking that, just because an EndTxn
claimed to have been able to commit a transaction, that means that any
WriteTooOldError encountered previously by the batch was safe to
discard. An EndTxn might consider that it can commit even if there had
been previous write too old conditions if the NoRefreshSpans flag is
set. The problems is that a CPut that had returned a WriteTooOldError
also evaluated at the wrong read timestamp, and so its evaluation can't
be relied on.

Another issue is that, when the EndTxn code mentioned above considers
that it's safe to commit at a bumped timestamp, it doesn't take into
considerations that the EndTxn's batch might have performed reads (other
than CPuts)  that have been evaluated at a lower timestamp. This can
happen, for example in the following scenario: - a txn sends a Put which
gets bumped by the ts cache - the txn then sends a Scan + EndTxn. The
scan gets evaluated at the original timestamp, but then we commit at a
bumped one because the NoRefreshSpans flag is set.

The patch fixes the bugs by reworking how evaluation takes advantage of
the fact that some requests have flexible timestamps. EndTxn no longer
is in the business of committing at bumped timestamps, and its code is
thus simplified. Instead, the replica's "local retries" loop takes over.
The replica already had code handling non-transactional batches that
evaluated them repeatedly in case of WriteTooOldErrors. This patch
rationalizes and expands this code to deal with transactional batches
too, and with pushes besides WriteTooOldErrors. This reevaluation loop
now handles the cases in which the EndTxn used to bump the commit
timestamp.

The patch also fixes a third bug: the logic evaluateBatch() for
resetting the WriteTooOld state after a successful EndTransaction was
ignoring the STAGING state, meaning that the server would return a
WriteTooOldError even though the transaction was committed. I'm not sure
if this had dramatic consequences or was benign...

Fixes #42849

Release note (bug fix): A bug causing lost update transaction anomalies
was fixed.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/42969)
<!-- Reviewable:end -->

andreimatei commented:
First commit is #42965

bdarnell reviewed (approved):
> Reviewed 5 of 5 files at r1, 9 of 9 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42969)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)



<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 1 of 5 files at r1, 8 of 9 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42969)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei)

___
*[pkg/kv/dist_sender_server_test.go, line 2603 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-LvTvK6ZCrqthmDyeNGW:-LvTvK6_5LUt4w9P4QDy:bf1u6xt) ([raw file](https://github.com/cockroachdb/cockroach/blob/41ef7397ae24c792ff6408a5c946c277963ca9b2/pkg/kv/dist_sender_server_test.go#L2603)):*
> ```Go
> 				b.CPut("a", "cput", strToValue("orig"))
> 				b.Put("c", "put")
> 				err := txn.CommitInBatch(ctx, b)
> ```

What's up with this?
___
*[pkg/kv/dist_sender_server_test.go, line 2642 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-LvTvMbA5GZV-3fipilB:-LvTvMbA5GZV-3fipilC:b-otzyl7) ([raw file](https://github.com/cockroachdb/cockroach/blob/41ef7397ae24c792ff6408a5c946c277963ca9b2/pkg/kv/dist_sender_server_test.go#L2642)):*
> ```Go
> 				return txn.CommitInBatch(ctx, b)
> 			},
> 			// Expect a TxcCoordSender retry, which should succeed.
> ```

TxnCoordSender
___
*[pkg/roachpb/api.proto, line 1806 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-LvTvRvw6koaoSIfzLjV:-LvTvRvw6koaoSIfzLjW:b-5jk741) ([raw file](https://github.com/cockroachdb/cockroach/blob/41ef7397ae24c792ff6408a5c946c277963ca9b2/pkg/roachpb/api.proto#L1806)):*
> ```protobuf
>   // nowhere to defer the error to). Similarly, it's pointless to set the flag
>   // on batches containing an EndTransaction - although we do allow it and it
>   // effects sub-batches split off by the DistSender.
> ```

Why do we do it if it's pointless? I would say either less here or more.
___
*[pkg/storage/replica_evaluate.go, line 369 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-Lvftv5aBtgyOst0xuP9:-Lvftv5aBtgyOst0xuPA:bh6opz1) ([raw file](https://github.com/cockroachdb/cockroach/blob/41ef7397ae24c792ff6408a5c946c277963ca9b2/pkg/storage/replica_evaluate.go#L369)):*
> ```Go
> 	// If there was an EndTransaction in the batch that finalized the transaction,
> 	// the WriteTooOld status has been fully processed and we can discard the error.
> 	if baHeader.Txn != nil && baHeader.Txn.Status.IsFinalizedOrStaging() {
> ```

Are there any tests that hit the `OrStaging` part of this condition?
___
*[pkg/storage/replica_evaluate.go, line 387 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-LvfuCWL30fE9zf051p6:-LvfuCWL30fE9zf051p7:b-7rm1iq) ([raw file](https://github.com/cockroachdb/cockroach/blob/41ef7397ae24c792ff6408a5c946c277963ca9b2/pkg/storage/replica_evaluate.go#L387)):*
> ```Go
> 		// retry (i.e. there were no refresh spans and the commit timestamp
> 		// wasn't leaked).
> 		if baHeader.Txn.Status == roachpb.COMMITTED {
> ```

Why do we need this? What fails without it? Should we also have it for transactions with a STAGING status?
___
*[pkg/storage/stores.go, line 175 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-Lvfv2W8BcIcESEm3ZDy:-Lvfv2W8BcIcESEm3ZDz:b-poebaw) ([raw file](https://github.com/cockroachdb/cockroach/blob/41ef7397ae24c792ff6408a5c946c277963ca9b2/pkg/storage/stores.go#L175)):*
> ```Go
> 		log.Fatal(ctx, "batch request missing store ID")
> 	}
> 	if ba.Header.DeferWriteTooOldError && ba.Txn == nil {
> ```

This is a pretty high-level code-path to be dropping such a specific condition into. Should we add a `BatchRequest.Validate` method instead? That can also check the conditions above.


<!-- Sent from Reviewable.io -->

andreimatei committed:
storage: rationalize updates to br.Timestamp

This patch fixes an incorrect comment on br.Timestamp that mistakenly
indicated that the field is only relevant for non-transactional
requests. It also makes the management of br.Timestamp in
evaluateBatch() - it used to be pretty unscrutable.

Release note: None

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42969)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @bdarnell, and @nvanbenschoten)

___
*[pkg/kv/dist_sender_server_test.go, line 2603 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-LvTvK6ZCrqthmDyeNGW:-Lw-gtel7EfFjP8XiAPb:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/41ef7397ae24c792ff6408a5c946c277963ca9b2/pkg/kv/dist_sender_server_test.go#L2603)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

What's up with this?
</blockquote></details>

done
___
*[pkg/kv/dist_sender_server_test.go, line 2642 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-LvTvMbA5GZV-3fipilB:-Lw-guj-1CBnViiPPsMb:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/41ef7397ae24c792ff6408a5c946c277963ca9b2/pkg/kv/dist_sender_server_test.go#L2642)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

TxnCoordSender
</blockquote></details>

done
___
*[pkg/roachpb/api.proto, line 1806 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-LvTvRvw6koaoSIfzLjV:-Lw-iZOQ7JGOq4xj8OkJ:b-x85is1) ([raw file](https://github.com/cockroachdb/cockroach/blob/41ef7397ae24c792ff6408a5c946c277963ca9b2/pkg/roachpb/api.proto#L1806)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Why do we do it if it's pointless? I would say either less here or more.
</blockquote></details>

See now. Also keep in mind that we don't really "do it", since this flag is never set :P
___
*[pkg/storage/replica_evaluate.go, line 369 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-Lvftv5aBtgyOst0xuP9:-Lw0EzSG-jgR2J8ktC6W:beph79p) ([raw file](https://github.com/cockroachdb/cockroach/blob/41ef7397ae24c792ff6408a5c946c277963ca9b2/pkg/storage/replica_evaluate.go#L369)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Are there any tests that hit the `OrStaging` part of this condition?
</blockquote></details>

I'll add a test
___
*[pkg/storage/replica_evaluate.go, line 387 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-LvfuCWL30fE9zf051p6:-Lw0F-vRB3igJqCdbm5I:b-tthttg) ([raw file](https://github.com/cockroachdb/cockroach/blob/41ef7397ae24c792ff6408a5c946c277963ca9b2/pkg/storage/replica_evaluate.go#L387)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Why do we need this? What fails without it? Should we also have it for transactions with a STAGING status?
</blockquote></details>

prepended a commit removing this. PTAL


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 2 of 11 files at r4, 9 of 9 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42969)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei)

___
*[pkg/roachpb/api.proto, line 1804 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-LwFgyCgDoKVV1ywT6RV:-LwFgyCgDoKVV1ywT6RW:b-v8cnnr) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf28026434e580bdb6bcae16dd6a94139296bc77/pkg/roachpb/api.proto#L1804)):*
> ```protobuf
>   //
>   // Non-transactional requests are not allowed to set this flag (since there's
>   // nowhere to defer the error to). At the replica level, this flag is ignore
> ```

s/ignore/ignored/
___
*[pkg/storage/replica_evaluate.go, line 369 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-Lvftv5aBtgyOst0xuP9:-LwFhLhN770As5lM-C8I:bd5ty59) ([raw file](https://github.com/cockroachdb/cockroach/blob/41ef7397ae24c792ff6408a5c946c277963ca9b2/pkg/storage/replica_evaluate.go#L369)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I'll add a test
</blockquote></details>

:eyes: 
___
*[pkg/storage/replica_write.go, line 437 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-LwFhCWVAUJrnGamVpPi:-LwFhCWVAUJrnGamVpPj:b2nglp8) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf28026434e580bdb6bcae16dd6a94139296bc77/pkg/storage/replica_write.go#L437)):*
> ```Go
> 				break
> 			}
> 			log.Infof(ctx, "!!! local retries updating ba.Timestamp %s->%s", ba.Timestamp, wtoErr.ActualTimestamp)
> ```

!!!
___
*[pkg/storage/stores.go, line 175 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-Lvfv2W8BcIcESEm3ZDy:-LwFhDecCjfn19foiQWx:b-8fznq3) ([raw file](https://github.com/cockroachdb/cockroach/blob/41ef7397ae24c792ff6408a5c946c277963ca9b2/pkg/storage/stores.go#L175)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This is a pretty high-level code-path to be dropping such a specific condition into. Should we add a `BatchRequest.Validate` method instead? That can also check the conditions above.
</blockquote></details>

I still think we should do something like this.


<!-- Sent from Reviewable.io -->

andreimatei committed:
roachpb: rename NoRefreshSpans -> CanCommitAtHigherTimestamp

Release note: None

andreimatei commented:
I've reworked this, PTAL.
Turned out we had yet another bug than the original CPut one - described in the commit msg.

One thing I'm unsure about is what the story around non-transactional batches and timestamp bumps is. What's supposed to happen, for example, to a non-txn batch consisting of a `Put+Scan`, where the `Put` is pushed by the timestamp cache? The `DistSender` will split the two, and then the `Scan` will evaluate at... which timestamp?

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42969)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell and @nvanbenschoten)

___
*[pkg/roachpb/api.proto, line 1804 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-LwFgyCgDoKVV1ywT6RV:-Lw_b56d6FCBvWyw3x6C:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf28026434e580bdb6bcae16dd6a94139296bc77/pkg/roachpb/api.proto#L1804)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

s/ignore/ignored/
</blockquote></details>

done
___
*[pkg/storage/replica_evaluate.go, line 369 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-Lvftv5aBtgyOst0xuP9:-LwUf8Kl1nVUg4zH3I4l:b2rz1ni) ([raw file](https://github.com/cockroachdb/cockroach/blob/41ef7397ae24c792ff6408a5c946c277963ca9b2/pkg/storage/replica_evaluate.go#L369)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

:eyes: 
</blockquote></details>

👀 now
___
*[pkg/storage/replica_write.go, line 437 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-LwFhCWVAUJrnGamVpPi:-LwUfAIrDSMi8gOFYtVt:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf28026434e580bdb6bcae16dd6a94139296bc77/pkg/storage/replica_write.go#L437)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

!!!
</blockquote></details>

done
___
*[pkg/storage/stores.go, line 175 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-Lvfv2W8BcIcESEm3ZDy:-LwUgKJg29fbtUwSLDxi:b-jikiha) ([raw file](https://github.com/cockroachdb/cockroach/blob/41ef7397ae24c792ff6408a5c946c277963ca9b2/pkg/storage/stores.go#L175)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I still think we should do something like this.
</blockquote></details>

my silence merely meant that I had not gotten to it. Done.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> One thing I'm unsure about is what the story around non-transactional batches and timestamp bumps is. What's supposed to happen, for example, to a non-txn batch consisting of a Put+Scan, where the Put is pushed by the timestamp cache? The DistSender will split the two, and then the Scan will evaluate at... which timestamp?


I can now answer my question: a write running into the ts cache does not affect the timestamp of "non-write" requests from the same batch because all the non-writes have been split off by the DistSender and their sub-batches are not updated to the bumped timestamp. `CPuts` are affected by the timestamp bump - so it's possible for a non-transactional batch to read some data at ts0 but execute a CPut at ts1. Seems funky, but we probably don't care about non-txn batches too much.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-:-Lx7j00K6vsm_GyxYciE:bbfanaj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell and @nvanbenschoten)



<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 9 of 9 files at r6, 13 of 13 files at r7, 15 of 15 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42969)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei)

___
*[pkg/kv/dist_sender_server_test.go, line 2457 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-Lx8JI-w6hqee1QJTVDT:-Lx8JI-w6hqee1QJTVDU:bh7xt1n) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ce8507e80539eef543f526f847de9dad97983ce/pkg/kv/dist_sender_server_test.go#L2457)):*
> ```Go
> 			},
> 			retryable: func(ctx context.Context, txn *client.Txn) error {
> 				if err := txn.Put(ctx, "aother", "another put"); err != nil {
> ```

s/aother/another/
___
*[pkg/kv/dist_sender_server_test.go, line 2463 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-Lx8JD796UuO5N1XpTCi:-Lx8JD796UuO5N1XpTCj:b-3m2xif) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ce8507e80539eef543f526f847de9dad97983ce/pkg/kv/dist_sender_server_test.go#L2463)):*
> ```Go
> 				b.Put("a", "final value")
> 				return txn.CommitInBatch(ctx, b)
> 			},
> ```

Add some comment here about why the parallel commit was allowed even though the timestamp was forwarded.
___
*[pkg/kv/txn_interceptor_span_refresher.go, line 168 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-Lx8Jmr82Fl85KXsfkTm:-Lx8Jmr82Fl85KXsfkTn:b-ieqpgf) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ce8507e80539eef543f526f847de9dad97983ce/pkg/kv/txn_interceptor_span_refresher.go#L168)):*
> ```Go
> 	br, pErr, largestRefreshTS := sr.sendLockedWithRefreshAttempts(ctx, ba, maxAttempts)
> 	if pErr != nil {
> 		// The server sometimes "performs refreshes" - it updates the transaction's
> ```

We only need to do this on the error path?
___
*[pkg/roachpb/batch.go, line 600 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-Lx8Jwnx3n-OfkRu0pQ0:-Lx8Jwnx3n-OfkRu0pQ1:b-fiqtex) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ce8507e80539eef543f526f847de9dad97983ce/pkg/roachpb/batch.go#L600)):*
> ```Go
> }
> 
> // ValidateForEvaluation performs sanity checks on the batch.
> ```

Note where this validation is meant to take place. E.g. this is server-side validation.
___
*[pkg/storage/replica_evaluate.go, line 369 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-Lvftv5aBtgyOst0xuP9:-Lx8IFzYEB_76FiSrDoM:bskm9bp) ([raw file](https://github.com/cockroachdb/cockroach/blob/41ef7397ae24c792ff6408a5c946c277963ca9b2/pkg/storage/replica_evaluate.go#L369)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

👀 now
</blockquote></details>

:eyes:'d. Very nice
___
*[pkg/storage/replica_evaluate.go, line 368 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-Lx8IU6E2B9Qx405bhx5:-Lx8IU6E2B9Qx405bhx6:b-5jkyf) ([raw file](https://github.com/cockroachdb/cockroach/blob/200d8df2009eb8261efe6924641f15c7279083d3/pkg/storage/replica_evaluate.go#L368)):*
> ```Go
> 		// EndTransaction that decided that it can essentially refresh to something
> 		// higher than baHeader.Timestamp because there were no refresh spans.
> 		br.Timestamp.Forward(baHeader.Txn.ReadTimestamp)
> ```

minor nit: I'd `s/baHeader.Txn.ReadTimestamp/br.Txn.ReadTimestamp/`. You just set `br.Txn` so that doesn't change any logic, but it seems a little closer to what you semantically mean here. It's not otherwise obvious that `baHeader.Txn` was being mutated along the way.
___
*[pkg/storage/replica_evaluate.go, line 335 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-Lx8KZzO1C-KVq9ocssz:-Lx8KZzO1C-KVq9ocst-:bkm47fs) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ce8507e80539eef543f526f847de9dad97983ce/pkg/storage/replica_evaluate.go#L335)):*
> ```Go
> 
> 				// If this write request also implies a read, it morally needs to be
> 				// "refreshed".
> ```

it morally needs to be "refreshed" if it hit a WriteTooOldError.

Just so we don't cause any confusion with the fact that `!roachpb.NeedsRefresh(args)` for these requests.
___
*[pkg/storage/replica_evaluate.go, line 446 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-Lx8Kx6GE6AosQkjDP9S:-Lx8Kx6GE6AosQkjDP9T:b-lwtwji) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ce8507e80539eef543f526f847de9dad97983ce/pkg/storage/replica_evaluate.go#L446)):*
> ```Go
> 	var pd result.Result
> 
> 	cArgs := batcheval.CommandArgs{
> ```

Did you mean to keep this diff?
___
*[pkg/storage/replica_write.go, line 399 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-Lx8LQOx9VZNTWBNGEvu:-Lx8LQOx9VZNTWBNGEvv:b-8ky48) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ce8507e80539eef543f526f847de9dad97983ce/pkg/storage/replica_write.go#L399)):*
> ```Go
> 
> 		br, res, pErr = evaluateBatch(ctx, idKey, batch, rec, ms, ba, false /* readOnly */)
> 		if pErr == nil {
> ```

nit: `if pErr == nil && opLogger != nil {`
___
*[pkg/storage/replica_write.go, line 416 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-Lx8MKnoDjAbYwqA3SNc:-Lx8MKnpBT5bjH2pamHf:blvafck) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ce8507e80539eef543f526f847de9dad97983ce/pkg/storage/replica_write.go#L416)):*
> ```Go
> }
> 
> func canDoServersideRetry(ctx context.Context, pErr *roachpb.Error, ba *roachpb.BatchRequest) bool {
> ```

I'd document this a bit. Also, make it clear that it can mutate `ba`.
___
*[pkg/storage/replica_write.go, line 417 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-Lx8LmH96uo6znXMg8jV:-Lx8LmH96uo6znXMg8jW:blgjb0p) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ce8507e80539eef543f526f847de9dad97983ce/pkg/storage/replica_write.go#L417)):*
> ```Go
> 
> func canDoServersideRetry(ctx context.Context, pErr *roachpb.Error, ba *roachpb.BatchRequest) bool {
> 	if pErr == nil {
> ```

It seems like there's some redundancy between checking `pErr` here and a few lines above. Also, is this method trying to determine whether we "can" do a server-side retry or whether we "should" do a server-side retry. Taking `pErr` into account here implies the latter.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42969)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[pkg/kv/dist_sender_server_test.go, line 2457 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-Lx8JI-w6hqee1QJTVDT:-LxCoH9J7U1jOc-lmZoq:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ce8507e80539eef543f526f847de9dad97983ce/pkg/kv/dist_sender_server_test.go#L2457)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

s/aother/another/
</blockquote></details>

done
___
*[pkg/kv/dist_sender_server_test.go, line 2463 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-Lx8JD796UuO5N1XpTCi:-LxCoYRPE9NTaXc4EmzD:bk5taww) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ce8507e80539eef543f526f847de9dad97983ce/pkg/kv/dist_sender_server_test.go#L2463)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Add some comment here about why the parallel commit was allowed even though the timestamp was forwarded.
</blockquote></details>

added more words above
___
*[pkg/kv/txn_interceptor_span_refresher.go, line 168 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-Lx8Jmr82Fl85KXsfkTm:-LxCod4pC4BINHXT3emg:b-y95mmw) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ce8507e80539eef543f526f847de9dad97983ce/pkg/kv/txn_interceptor_span_refresher.go#L168)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We only need to do this on the error path?
</blockquote></details>

Yes. We could do it on success too if you prefer, at the expense of more lines of code.
As I'm trying to explain in the comment, the server does these refreshes only on commits. If the commit succeeded, there's nothing left for the client to do, so the state updates don't matter.
If the commit fails, the state does matter but only because of a pretty weak reason. That's why I wanted to keep the code changes to a minimum.
___
*[pkg/roachpb/batch.go, line 600 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-Lx8Jwnx3n-OfkRu0pQ0:-LxCphy22YOc60q36tj3:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ce8507e80539eef543f526f847de9dad97983ce/pkg/roachpb/batch.go#L600)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Note where this validation is meant to take place. E.g. this is server-side validation.
</blockquote></details>

done
___
*[pkg/storage/replica_evaluate.go, line 368 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-Lx8IU6E2B9Qx405bhx5:-LxCqSgH7Hj9azJm6OXn:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/200d8df2009eb8261efe6924641f15c7279083d3/pkg/storage/replica_evaluate.go#L368)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

minor nit: I'd `s/baHeader.Txn.ReadTimestamp/br.Txn.ReadTimestamp/`. You just set `br.Txn` so that doesn't change any logic, but it seems a little closer to what you semantically mean here. It's not otherwise obvious that `baHeader.Txn` was being mutated along the way.
</blockquote></details>

done
___
*[pkg/storage/replica_evaluate.go, line 335 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-Lx8KZzO1C-KVq9ocssz:-LxCq6DU1jDi4REncMei:bqkyrdm) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ce8507e80539eef543f526f847de9dad97983ce/pkg/storage/replica_evaluate.go#L335)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

it morally needs to be "refreshed" if it hit a WriteTooOldError.

Just so we don't cause any confusion with the fact that `!roachpb.NeedsRefresh(args)` for these requests.
</blockquote></details>

you meant add more words to the comment, right? Done.
___
*[pkg/storage/replica_evaluate.go, line 446 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-Lx8Kx6GE6AosQkjDP9S:-LxCqetN1DYFLZzE3QDl:bkf6yt3) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ce8507e80539eef543f526f847de9dad97983ce/pkg/storage/replica_evaluate.go#L446)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Did you mean to keep this diff?
</blockquote></details>

no. gone.
___
*[pkg/storage/replica_write.go, line 399 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-Lx8LQOx9VZNTWBNGEvu:-LxCr-WB0_9StNVSsATa:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ce8507e80539eef543f526f847de9dad97983ce/pkg/storage/replica_write.go#L399)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: `if pErr == nil && opLogger != nil {`
</blockquote></details>

done
___
*[pkg/storage/replica_write.go, line 416 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-Lx8MKnoDjAbYwqA3SNc:-LxCsbtx4v_kERchACJ-:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ce8507e80539eef543f526f847de9dad97983ce/pkg/storage/replica_write.go#L416)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I'd document this a bit. Also, make it clear that it can mutate `ba`.
</blockquote></details>

done
___
*[pkg/storage/replica_write.go, line 417 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-Lx8LmH96uo6znXMg8jV:-LxCu__4ELtyYan-scBt:bv36g0w) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ce8507e80539eef543f526f847de9dad97983ce/pkg/storage/replica_write.go#L417)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

It seems like there's some redundancy between checking `pErr` here and a few lines above. Also, is this method trying to determine whether we "can" do a server-side retry or whether we "should" do a server-side retry. Taking `pErr` into account here implies the latter.
</blockquote></details>

Removed the check for `pErr`.


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 7 of 13 files at r10, 15 of 15 files at r11.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-:-LxwEBiLE3JA4mcnLSMd:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei)

___
*[pkg/kv/txn_interceptor_span_refresher.go, line 168 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-Lx8Jmr82Fl85KXsfkTm:-LxwDdFs5pJJdDiNCN6Y:bb6sh2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ce8507e80539eef543f526f847de9dad97983ce/pkg/kv/txn_interceptor_span_refresher.go#L168)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Yes. We could do it on success too if you prefer, at the expense of more lines of code.
As I'm trying to explain in the comment, the server does these refreshes only on commits. If the commit succeeded, there's nothing left for the client to do, so the state updates don't matter.
If the commit fails, the state does matter but only because of a pretty weak reason. That's why I wanted to keep the code changes to a minimum.
</blockquote></details>

Ack.
___
*[pkg/kv/txn_interceptor_span_refresher.go, line 171 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-LxwDZob0brqPSVJf3u6:-LxwDZob0brqPSVJf3u7:brhfpmw) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2ffe714e1a0001cf00a8da6efdcc67a503198db/pkg/kv/txn_interceptor_span_refresher.go#L171)):*
> ```Go
> 		// ReadTimestamp when the client said that there's nothing to refresh. In
> 		// these cases, we need to update our refreshedTimestamp too. This is pretty
> 		// inconsequential: the server only does this on an EndTransaction. If the
> ```

s/EndTransaction/EndTxn/

Do you mind searching this entire diff for that change?


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42969)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/kv/txn_interceptor_span_refresher.go, line 171 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-LxwDZob0brqPSVJf3u6:-LxwJkwv8h0mDIZ-26Re:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2ffe714e1a0001cf00a8da6efdcc67a503198db/pkg/kv/txn_interceptor_span_refresher.go#L171)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

s/EndTransaction/EndTxn/

Do you mind searching this entire diff for that change?
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

andreimatei commented:
I've reworked the patch again to take EndTxn out of the bumping timestamps business, which takes the patch to a new local maximum. See updated commit message.
PTAL

andreimatei reviewed (commented):
r12 is https://reviewable.io/reviews/cockroachdb/cockroach/43791

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-:-Ly0xlKa9IDdunQvrPKM:b-nw7me2)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @nvanbenschoten)



<!-- Sent from Reviewable.io -->

andreimatei committed:
roachpb: improve comments on the write_too_old field

Explain that it's about the quality of the serializability error that
is produced.

Release note: None

nvb reviewed (commented):
> Reviewed 8 of 14 files at r14, 17 of 17 files at r15, 2 of 2 files at r16.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42969)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andreimatei)

___
*[pkg/roachpb/batch.go, line 490 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-LyBDO557D8Gt2g0P-hu:-LyBDO557D8Gt2g0P-hv:b-eiq5y6) ([raw file](https://github.com/cockroachdb/cockroach/blob/8089a9418e2f3627aab054cc07537a3f617073ba/pkg/roachpb/batch.go#L490)):*
> ```Go
> // can sometimes have their read timestamp bumped on the server, which doesn't
> // work for read requests due to have the timestamp-aware latching works (i.e. a
> // read that acquired a latch @ ts10 can't simply be bumped to tsrequests 20
> ```

"tsrequests"
___
*[pkg/storage/replica_evaluate.go, line 368 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-Lx8IU6E2B9Qx405bhx5:-LyBET6ADMrjMD4kfqUh:bgwx2ud) ([raw file](https://github.com/cockroachdb/cockroach/blob/200d8df2009eb8261efe6924641f15c7279083d3/pkg/storage/replica_evaluate.go#L368)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

done
</blockquote></details>

Did we regress on this?
___
*[pkg/storage/replica_evaluate.go, line 312 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-LyBE6NQE_ArSKqWiub0:-LyBE6NQE_ArSKqWiub1:b-9s9bty) ([raw file](https://github.com/cockroachdb/cockroach/blob/8089a9418e2f3627aab054cc07537a3f617073ba/pkg/storage/replica_evaluate.go#L312)):*
> ```Go
> 				if baHeader.Txn != nil {
> 					baHeader.Txn.WriteTimestamp.Forward(tErr.ActualTimestamp)
> 					log.VEventf(ctx, 2, "setting WTO. idx: %d, req: %s", index, args)
> ```

Are you meaning to keep this? If so, we should make it more readable.
___
*[pkg/storage/replica_write.go, line 260 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-LyBG3KoB_EgKhrCY9rg:-LyBG3KoB_EgKhrCY9rh:b-az2ym3) ([raw file](https://github.com/cockroachdb/cockroach/blob/8089a9418e2f3627aab054cc07537a3f617073ba/pkg/storage/replica_write.go#L260)):*
> ```Go
> 	// timestamp, let's evaluate the batch at the bumped timestamp. This will
> 	// allow it commit, and also it'll allow us to attempt the 1PC code path.
> 	maybeBumpReadTimestampToWriteTimestamp(ctx, ba)
> ```

What do you think about merging some of this logic with `maybeStripInFlightWrites` and creating a more general server-side request update path? The reason for all of these kinds of updates is that we need to wait until we're routed to a Replica before we're able to make certain decisions. It seems like that kind of logic should then be centralized.
___
*[pkg/storage/replica_write.go, line 300 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-LyBGmyQ4DWF8Zd34RcJ:-LyBGmyR0tEMStDROPiY:bw2zofp) ([raw file](https://github.com/cockroachdb/cockroach/blob/8089a9418e2f3627aab054cc07537a3f617073ba/pkg/storage/replica_write.go#L300)):*
> ```Go
> 		synthesizeEndTxnResponse := func() (_res onePCResult) {
> 			// Handle the case of a required one phase commit transaction.
> 			defer func() {
> ```

Why does this need to be a defer-ed function? This control flow seems much more difficult to understand than before. What do we gain from it?
___
*[pkg/storage/replica_write.go, line 332 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-LyBGFzN-V1ie4CT_VTV:-LyBGFzN-V1ie4CT_VTW:b-sjhzrh) ([raw file](https://github.com/cockroachdb/cockroach/blob/8089a9418e2f3627aab054cc07537a3f617073ba/pkg/storage/replica_write.go#L332)):*
> ```Go
> 			}
> 
> 			// 1PC execution was successful, let's synthesize a EndTxnResponse.
> ```

s/a/an/
___
*[pkg/storage/replica_write.go, line 481 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-LyBHJhe9Dh0j-xTBhGK:-LyBHJhe9Dh0j-xTBhGL:bl52n2k) ([raw file](https://github.com/cockroachdb/cockroach/blob/8089a9418e2f3627aab054cc07537a3f617073ba/pkg/storage/replica_write.go#L481)):*
> ```Go
> 
> func canDoServersideRetry(ctx context.Context, pErr *roachpb.Error, ba *roachpb.BatchRequest) bool {
> 	if pErr == nil {
> ```

I thought we had removed this.
___
*[pkg/storage/replica_write.go, line 526 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-LyBFgv5BQ7xt_wwgxTK:-LyBFgv5BQ7xt_wwgxTL:b-ln10vx) ([raw file](https://github.com/cockroachdb/cockroach/blob/8089a9418e2f3627aab054cc07537a3f617073ba/pkg/storage/replica_write.go#L526)):*
> ```Go
> // Note that this, like all the server-side bumping of the read timestamp, only
> // works for batches that exclusively contain writes; reads cannot be bumped
> // liked this because they've already acquired timestamp-aware latches.
> ```

"bumped liked this"
___
*[pkg/storage/replica_write.go, line 548 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-LyBFQ0uAW-0sit4nk3C:-LyBFQ0uAW-0sit4nk3D:b-qr8jm5) ([raw file](https://github.com/cockroachdb/cockroach/blob/8089a9418e2f3627aab054cc07537a3f617073ba/pkg/storage/replica_write.go#L548)):*
> ```Go
> 				"ba.Txn.WriteTimestamp: %s", ts, txn.ReadTimestamp, txn.WriteTimestamp)
> 		}
> 		txn.ReadTimestamp = ts
> ```

This should be copy-on-write, right?


<!-- Sent from Reviewable.io -->

andreimatei committed:
storage: move maybeStripInFlightWrites

This patch creates a dedicated file for functions performing server-side
modifications to a batch. The next commit will have a 2nd such function.

Release note: None

andreimatei committed:
storage: simplify TestIsOnePhaseCommit

This test was creating batches with different flags set. The combination
of the WriteTooOld flag being set and the write timestamp not being
bumped in relation to the read timestamp does not make sense. All such
tests are removed, and the test spec is improved.
A future commit introduces an assertion that such requests are indeed
not received by servers.

Release note: None

andreimatei reviewed (commented):
I needed to prepend r17.
Also please take a look at the deleted tests in cmd_end_transaction_test.go and tell me if you feel like something else should have been done about them - `EndTxn` itself no longer looks at the `CanCommitAtHigherTimestamp` flag, and so all those tests now return retriable errors that are swallowed by the replica retry loop.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-:-LyWLGdi-HUwGKwRWNID:b-26zyvx)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @nvanbenschoten)

___
*[pkg/roachpb/batch.go, line 490 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-LyBDO557D8Gt2g0P-hu:-LyVHpbfBg5n4Ew5pX0Q:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/8089a9418e2f3627aab054cc07537a3f617073ba/pkg/roachpb/batch.go#L490)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

"tsrequests"
</blockquote></details>

done
___
*[pkg/storage/replica_evaluate.go, line 368 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-Lx8IU6E2B9Qx405bhx5:-LyVJZX25MhJjAUVthJN:b-hyjysw) ([raw file](https://github.com/cockroachdb/cockroach/blob/200d8df2009eb8261efe6924641f15c7279083d3/pkg/storage/replica_evaluate.go#L368)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Did we regress on this?
</blockquote></details>

i guess so, I don't know what happened. Fixed.
___
*[pkg/storage/replica_evaluate.go, line 312 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-LyBE6NQE_ArSKqWiub0:-LyVIV95E8e63--iz-Rf:b-hfpv0l) ([raw file](https://github.com/cockroachdb/cockroach/blob/8089a9418e2f3627aab054cc07537a3f617073ba/pkg/storage/replica_evaluate.go#L312)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Are you meaning to keep this? If so, we should make it more readable.
</blockquote></details>

removed
___
*[pkg/storage/replica_write.go, line 260 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-LyBG3KoB_EgKhrCY9rg:-LyVPnGV2SD2j4MbX6zl:b-irdabs) ([raw file](https://github.com/cockroachdb/cockroach/blob/8089a9418e2f3627aab054cc07537a3f617073ba/pkg/storage/replica_write.go#L260)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

What do you think about merging some of this logic with `maybeStripInFlightWrites` and creating a more general server-side request update path? The reason for all of these kinds of updates is that we need to wait until we're routed to a Replica before we're able to make certain decisions. It seems like that kind of logic should then be centralized.
</blockquote></details>

As discussed, they unfortunately can't be called together, but I've moved them both to a new, dedicated, file.
___
*[pkg/storage/replica_write.go, line 300 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-LyBGmyQ4DWF8Zd34RcJ:-LyVjZmA8voPrDbiLPJN:bb728a) ([raw file](https://github.com/cockroachdb/cockroach/blob/8089a9418e2f3627aab054cc07537a3f617073ba/pkg/storage/replica_write.go#L300)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Why does this need to be a defer-ed function? This control flow seems much more difficult to understand than before. What do we gain from it?
</blockquote></details>

ok see now
___
*[pkg/storage/replica_write.go, line 332 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-LyBGFzN-V1ie4CT_VTV:-LyVhuA_8CDhXn8JVwkc:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/8089a9418e2f3627aab054cc07537a3f617073ba/pkg/storage/replica_write.go#L332)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

s/a/an/
</blockquote></details>

done
___
*[pkg/storage/replica_write.go, line 481 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-LyBHJhe9Dh0j-xTBhGK:-LyVkaES0SQ-qR_b1Hf5:b-abeet8) ([raw file](https://github.com/cockroachdb/cockroach/blob/8089a9418e2f3627aab054cc07537a3f617073ba/pkg/storage/replica_write.go#L481)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I thought we had removed this.
</blockquote></details>

Done. Not sure what happened, but I went through and re-incorporated a couple of small changes from the previous review.
___
*[pkg/storage/replica_write.go, line 526 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-LyBFgv5BQ7xt_wwgxTK:-LyVjsjq6j16EEQybhsK:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/8089a9418e2f3627aab054cc07537a3f617073ba/pkg/storage/replica_write.go#L526)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

"bumped liked this"
</blockquote></details>

done
___
*[pkg/storage/replica_write.go, line 548 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-LyBFQ0uAW-0sit4nk3C:-LyVkOc_9RVi1-iIg1Cd:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/8089a9418e2f3627aab054cc07537a3f617073ba/pkg/storage/replica_write.go#L548)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This should be copy-on-write, right?
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 11 of 27 files at r17, 7 of 14 files at r19, 1 of 4 files at r20, 19 of 19 files at r21, 2 of 2 files at r22.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-:-Lya2Y21DCqWPJ1NEJs0:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei)

___
*[pkg/storage/replica_batch_updates.go, line 26 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-Lya1Si16rtQ06bKYteD:-Lya1Si16rtQ06bKYteE:b-pb18ck) ([raw file](https://github.com/cockroachdb/cockroach/blob/b08c3e249b5c79c9225c560cbfb5df38854ede9e/pkg/storage/replica_batch_updates.go#L26)):*
> ```Go
> // This files contains functions performing updates to a BatchRequest performed
> // on the server-side, specifically after the request has been routed to a
> // replica.
> ```

:thumbsup: 

Could you make a note about both why these updates exist (e.g. they allow the batch to be updated based on range split boundaries) and also make a note that all of these methods should be copy-on-write?
___
*[pkg/storage/replica_test.go, line 505 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-Lya1xIu5jM5A7YlVEln:-Lya1xIu5jM5A7YlVElo:b2nglp8) ([raw file](https://github.com/cockroachdb/cockroach/blob/b08c3e249b5c79c9225c560cbfb5df38854ede9e/pkg/storage/replica_test.go#L505)):*
> ```Go
> 		{ru: txnReqsNoRefresh, isTxn: true, exp1PC: true},
> 		{ru: txnReqsNoRefresh, isTxn: true, isTSOff: true, exp1PC: true},
> 		// !!!
> ```

!!!
___
*[pkg/storage/batcheval/cmd_end_transaction_test.go, line 304 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-Lya333oCadQgm4SR_l2:-Lya333oCadQgm4SR_l3:b8ekabo) ([raw file](https://github.com/cockroachdb/cockroach/blob/811f75a0e24ce29607765f9deecd7bcf35fbeb11/pkg/storage/batcheval/cmd_end_transaction_test.go#L304)):*
> ```Go
> 			commit:         true,
> 			inFlightWrites: writes,
> 			noRefreshSpans: true,
> ```

Let's remove `noRefreshSpans` if no test case sets it to true anymore and it's not considered during EndTxn evaluation.


<!-- Sent from Reviewable.io -->

andreimatei committed:
roachpb: fix txn.Update() commutativity

Updates to the WriteTooOld field were not commutative. This patch fixes
that, by clarifying that the transaction with the higher ReadTimestamp
gets to dictate the WriteTooOld value.
I'm not sure what consequences this used to have, besides allowing for
the confusing case where the server would receive a request with the
WriteTooOld flag set, but with the ReadTimestamp==WriteTimestamp.  A
future commit introduces a sanity assertion that all the requests with
the WTO flag have a bumped WriteTimestamp.

Release note: None

andreimatei reviewed (commented):
As discussed, I've prepended a commit fixing `txn.Update()`. PTAL
Everything else is unchanged, modulo the minor edits addressing your last comments.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-:-LyaeUfI3JUH89KLERuS:blv505d)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @nvanbenschoten)

___
*[pkg/storage/replica_batch_updates.go, line 26 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-Lya1Si16rtQ06bKYteD:-LyaOmWd8xqjIhNCKcOa:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/b08c3e249b5c79c9225c560cbfb5df38854ede9e/pkg/storage/replica_batch_updates.go#L26)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

:thumbsup: 

Could you make a note about both why these updates exist (e.g. they allow the batch to be updated based on range split boundaries) and also make a note that all of these methods should be copy-on-write?
</blockquote></details>

done
___
*[pkg/storage/replica_test.go, line 505 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-Lya1xIu5jM5A7YlVEln:-LyaNKI_0gaQ88oKTHIZ:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/b08c3e249b5c79c9225c560cbfb5df38854ede9e/pkg/storage/replica_test.go#L505)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

!!!
</blockquote></details>

done
___
*[pkg/storage/batcheval/cmd_end_transaction_test.go, line 304 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-Lya333oCadQgm4SR_l2:-LyaNaY74SgyZPRhj5Vx:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/811f75a0e24ce29607765f9deecd7bcf35fbeb11/pkg/storage/batcheval/cmd_end_transaction_test.go#L304)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Let's remove `noRefreshSpans` if no test case sets it to true anymore and it's not considered during EndTxn evaluation.
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 30 files at r23, 7 of 14 files at r26, 1 of 5 files at r27, 19 of 19 files at r28, 2 of 2 files at r29.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-:-LyekGr45voCmCawJtIr:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42969#-:-LyekjOB3eZ5VjN2gfNK:b-ydi9ay)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->
